rust/routee-compass-powertrain/src/routee/vehicle/default/dual_fuel_vehicle.rs

```diff
@@ -0,0 +1,365 @@
+            battery_energy_unit: self.battery_energy_unit,
+        };
+
+        Ok(Arc::new(new_phev))
+    }
+}
+
+fn update_state(
+    state: &[StateVar],
+    electrical_energy: Energy,
+    gasoline_energy: Energy,
+) -> TraversalState {
+    let mut updated_state = Vec::with_capacity(state.len());
+
+    // accumulated electrical energy
+    updated_state.push(state[0] + electrical_energy.into());
+
+    // accumulated fuel energy
+    updated_state.push(state[1] + gasoline_energy.into());
+
+    // remaining battery energy
+    let current_battery_energy = get_remaining_battery_energy_from_state(state);
+    let new_battery_energy = (current_battery_energy - electrical_energy).max(Energy::new(0.0));
+    updated_state.push(new_battery_energy.into());
+
+    updated_state
+}
+fn get_electrical_energy_from_state(state: &[StateVar]) -> Energy {
+    Energy::new(state[0].0)
+}
+
+fn get_gasoline_energy_from_state(state: &[StateVar]) -> Energy {
+    Energy::new(state[1].0)
+}
+
+fn get_remaining_battery_energy_from_state(state: &[StateVar]) -> Energy {
+    Energy::new(state[2].0)
+}
+
+fn get_battery_soc_percent(vehicle: &DualFuelVehicle, state: &[StateVar]) -> f64 {
+    let battery_energy_unit = vehicle.battery_energy_unit;
+
+    let battery_capacity_kwh =
+        battery_energy_unit.convert(vehicle.battery_capacity, EnergyUnit::KilowattHours);
+
+    let remaining_battery_energy = get_remaining_battery_energy_from_state(state);
+
+    let remaining_battery_energy_kwh =
+        battery_energy_unit.convert(remaining_battery_energy, EnergyUnit::KilowattHours);
+
+    (remaining_battery_energy_kwh.as_f64() / battery_capacity_kwh.as_f64()) * 100.0
+}
+
+/// Compute the energy for the PHEV by converting gasoline to kWh.
+/// This uses a simplified operation in which we assume that if the battery
+/// SOC is greater than zero we can just operate on battery to traverse a link.
+/// This is not entirely realistic as it's possible to arrive at a link with
+/// 0.001% SOC and still need to use gasoline to traverse the link.
+///
+/// In the future we could make this more sophisticated by calculating
+/// the energy required to traverse the link using the battery and then
+/// finding the point at which we would have to switch to gasoline
+///
+/// Returns a tuple of (electrical_energy, electrical_energy_unit, gasoline_energy, gasoline_energy_unit)
+fn get_phev_energy(
+    vehicle: &DualFuelVehicle,
+    battery_soc_percent: f64,
+    speed: (Speed, SpeedUnit),
+    grade: (Grade, GradeUnit),
+    distance: (Distance, DistanceUnit),
+) -> Result<(Energy, EnergyUnit, Energy, EnergyUnit), TraversalModelError> {
+    let electrical_energy_unit = vehicle
+        .charge_depleting_model
+        .energy_rate_unit
+        .associated_energy_unit();
+    let gasoline_energy_unit = vehicle
+        .charge_sustain_model
+        .energy_rate_unit
+        .associated_energy_unit();
+
+    if battery_soc_percent > 0.0 {
```

Comment:
> assume we can just use the battery

i think logic like this would avoid the need to make that assumption:

```python
if soc == 0:
  energy = 0.0
  use_gas()
else:
  energy = predict() # use battery
  if energy < 0:
    energy = 0.0
    use_gas()
  else:
    use_battery()
```

Reply:
Yeah that's a good point, I think I'll add this into a new issue to upgrade the logic of the PHEV in general to be more accurate (including doing things like spitting electric and gas on the same link if we cross the battery threshold during that traversal).